lib/idn.c

```diff
@@ -157,52 +173,68 @@ static char *idn_decode(const char *input)
          compatibility */
       rc = IDN2_LOOKUP(input, &decoded, IDN2_TRANSITIONAL);
     if(rc != IDN2_OK)
-      decoded = NULL;
+      result = CURLE_URL_MALFORMAT;
   }
 #elif defined(USE_WIN32_IDN)
-  if(!Curl_win32_idn_to_ascii(input, &decoded))
-    decoded = NULL;
+  result = win32_idn_to_ascii(input, &decoded);
 #endif
-  return decoded;
+  if(!result)
+    *output = decoded;
+  return result;
 }
 
-static char *idn_encode(const char *puny)
+static CURLcode idn_encode(const char *puny, char **output)
 {
   char *enc = NULL;
 #ifdef USE_LIBIDN2
   int rc = idn2_to_unicode_8z8z(puny, &enc, 0);
   if(rc != IDNA_SUCCESS)
-    return NULL;
+    return rc == IDNA_MALLOC_ERROR ? CURLE_OUT_OF_MEMORY : CURLE_URL_MALFORMAT;
 #elif defined(USE_WIN32_IDN)
-  enc = Curl_win32_ascii_to_idn(puny);
+  CURLcode result = win32_ascii_to_idn(puny, &enc);
+  if(result)
+    return result;
 #endif
-  return enc;
+  *output = enc;
+  return CURLE_OK;
 }
 
-char *Curl_idn_decode(const char *input)
+CURLcode Curl_idn_decode(const char *input, char **output)
 {
-  char *d = idn_decode(input);
+  char *d;
+  CURLcode result = idn_decode(input, &d);
 #ifdef USE_LIBIDN2
   if(d) {
     char *c = strdup(d);
     idn2_free(d);
-    d = c;
+    if(c)
+      d = c;
+    else
+      result = CURLE_OUT_OF_MEMORY;
   }
 #endif
-  return d;
+  if(!result)
+    *output = d;
+  return result;
 }
 
-char *Curl_idn_encode(const char *puny)
+CURLcode Curl_idn_encode(const char *puny, char **output)
 {
-  char *d = idn_encode(puny);
+  char *d;
```

Comment:
This either needs to be declared as NULL or the if statement on line 226 should be checking the result. Currently on a failed idn conversion this causes a `free(): invalid pointer` error on my machine. Changing the if statement or declaring d as NULL fixed it for me

Reply:
:+1: